### PR TITLE
Concolic Tracing

### PR DESCRIPTION
**Pull request by julihoh**

### CODE REVIEW

File: libafl/src/bolts/shmem.rs
Comment (julihoh) [reply libafl/src/bolts/shmem.rs:984]: Looking at `core::fmt::Write`, I'm not sure it would be a good fit. It seems like it's just for formatting strings?

I don't know why the `Write` trait would required std in the first place, but there is not much I can do here I think...

Comment (domenukk) [reply libafl/src/bolts/shmem.rs:984]: Ah yes you're right it's for unicode stuff, you're using it for bytes, I guess?
Still, the Cursor itself wouldn't need to be `std`

**Comment by julihoh** [comment]
Note: I added some VS Code development container files without the intention to merge them later. Because of reasons, I need them in the repository for a short while, I hope this is fine for the time being.

File: libafl/src/bolts/shmem.rs
Comment (julihoh) [reply libafl/src/bolts/shmem.rs]: I removed the std requirement for the struct itself btw :)

File: .devcontainer/Dockerfile
Comment (domenukk) [inline .devcontainer/Dockerfile:1]: yeah I don't think this belongs in git ;)

File: .gitmodules
Comment (domenukk) [inline .gitmodules:1]: Should we eventually pull this in via Cargo? Or will that not work because it needs to be built using cmake?

File: libafl/src/bolts/serdeany.rs
Comment (domenukk) [inline libafl/src/bolts/serdeany.rs:39]: Can't you impl `SerdeAny` for `Box<xyz>` or something like that? Does it really need its own method?

File: libafl/src/bolts/shmem.rs
Comment (domenukk) [inline libafl/src/bolts/shmem.rs:991]: I guess this could `impl From for ShMemCursor`, right? Not sure if it would be better though

File: libafl_concolic/test/basic_concolic_fuzzer/src/command_executor.rs
Comment (domenukk) [inline libafl_concolic/test/basic_concolic_fuzzer/src/command_executor.rs:99]: Do we really need busywaiting here? Why not just `wait`?

File: libafl_concolic/test/smoke_test.sh
Comment (domenukk) [inline libafl_concolic/test/smoke_test.sh:20]: Want to add a github actions thingy for that?

**Review by domenukk — COMMENTED** [review]
Work of art <3

File: .devcontainer/Dockerfile
Comment (julihoh) [reply .devcontainer/Dockerfile:1]: agreed... I need it for my weird setup at the moment. I promise to remove it before merge though 🤞

File: .gitmodules
Comment (julihoh) [reply .gitmodules:1]: I am honestly not sure how exactly this should _ideally_ work. I'll need to gather some info an write up a more detailed explanation for a proper discussion soon. the short version is this: when you compile symcc, the executables compiled with symcc are (dynamically) linked against the symcc runtime (which we implement partially in rust).  this 'links' a symcc build to its runtime, which makes things a bit complicated.

In the end, it would be awesome if we could have something like afl_cc, just with symcc. so essentially, a build_script type crate that bundles symcc so that the user can build an instrumented binary simply using a build script like in the example fuzzers. however, since the symcc build and runtime are linked, this compiler crate would also depend on the implementation of the runtime, which, _ideally_ would be implemented by the user in a separate crate and pulled into the compiler wrapper crate. at this level of decency mess my head explodes a little, which is why I haven't come up with a proper plan yet...

as for right now: I'd like to keep it as a submodule as it makes development easy.

File: libafl/src/bolts/serdeany.rs
Comment (julihoh) [reply libafl/src/bolts/serdeany.rs:39]: > Can't you impl `SerdeAny` for `Box<xyz>` or something like that? Does it really need its own method?

not without losing the box.

I needed this implementation for the serdeanymap::remove that I implemented, but I don't actually use it anymore. I can remove it again. 

I ended up with this solution after reading this article: https://articles.bchlr.de/traits-dynamic-dispatch-upcasting . I just tried factoring this method out into an extension trait, but couldn't get it to work. it may be possible, but I'd rather not spend more time on this type system shenanigans. what speaks against the method though? its only implemented via macros anyways, so the user doesn't need to care.

File: libafl/src/bolts/shmem.rs
Comment (julihoh) [reply libafl/src/bolts/shmem.rs:991]: true. I'm not sure what's the "official" style guide here. when should from be implemented?

for reference: std::io::cursor [implements just a single](https://doc.rust-lang.org/std/io/struct.Cursor.html#method.new) "new" method with a single parameter and no from/into at all. So I suppose I should rename the method to simply "new"?

File: libafl_concolic/test/basic_concolic_fuzzer/src/command_executor.rs
Comment (julihoh) [reply libafl_concolic/test/basic_concolic_fuzzer/src/command_executor.rs:99]: In order to return early if the timeout has passed. This is actually my favourite pet peeve with the std::process::Command api: there is no way to specify a timeout for the execution. 

but this reminds me that I should probably kill the process on timeout

File: libafl_concolic/test/smoke_test.sh
Comment (julihoh) [reply libafl_concolic/test/smoke_test.sh:20]: yes this should be the ultimate goal, but I don't want to do it until the structure of where which component resides in the end is final, because I expected there will be quite a bit of changes.

File: libafl/src/bolts/shmem.rs
Comment (domenukk) [reply libafl/src/bolts/shmem.rs:991]: Makes sense to mimick that behavior, then(?)

Comment (julihoh) [reply libafl/src/bolts/shmem.rs]: yes :) I also changed it in a few other places. seems like I have tendency to call these methods `from_*` instead of `new`...

Comment (domenukk) [reply libafl/src/bolts/shmem.rs]: If you want a `from`, just implement the `From` trait, then rust will do the conversation for you with `blubber.into()`

**Comment by julihoh** [comment]
Thanks, much appreciated :)

**Review by domenukk — APPROVED** [review]

File: Dockerfile
Comment (julihoh) [reply Dockerfile:60]: Ah damnit. It should be a `main.rs`, right?

File: fuzzers/libfuzzer_stb_image_concolic/fuzzer/build.rs
Comment (julihoh) [reply fuzzers/libfuzzer_stb_image_concolic/fuzzer/build.rs:82]: no... missed the refactor here...

File: libafl_concolic/test/runtime_test/src/lib.rs
Comment (julihoh) [reply libafl_concolic/test/runtime_test/src/lib.rs:1]: this would make everything depend of SymCC (and the ability to build it, which depends on LLVM and stuff). This way, its all separate, so it can be tested and built only on targets that actually support it

**Comment by julihoh** [comment]
Okay, I addressed most feedback in the code, others I replied to.

regarding the eventual merge: I'd suggest to squash merge to keep the git history sane. right now, this branch contains a bunch of stuff that could turn out to be a pain later (like .devcontainer stuff and git submodules)

File: fuzzers/libfuzzer_stb_image_concolic/fuzzer/src/main.rs
Comment (julihoh) [reply fuzzers/libfuzzer_stb_image_concolic/fuzzer/src/main.rs:208]: It's not possible to do this as a mutator. We had discussed this at some point and came to the conclusion that the best solution would be to have a separate stage. I think the main issue was that concolic mutation produces an indeterminate amount of mutations, which does not fit with the current mutator API (which I think can only generate a fixed amount of mutations).

File: libafl/src/executors/command.rs
Comment (julihoh) [reply libafl/src/executors/command.rs:54]: I'll address this in my next PR, where I add a bunch of documentation in general.

File: libafl_concolic/symcc_runtime/build.rs
Comment (julihoh) [reply libafl_concolic/symcc_runtime/build.rs:18]: I honestly believe it's not even _that_ wrong :D

**Comment by julihoh** [comment]
Okay, I'd be ready to merge once we clear out the remaining unaddressed feedback. I've already got my next PR lined up with a bunch more documentation and small refactors, so I'd like to get this merged soon :)

File: Dockerfile
Comment (domenukk) [reply Dockerfile:60]: The idea in the dockerfile is currently:
copy a dummy lib, install dependencies, then copy the actual lib over later
that way, changing the lib.rs won't trigger a full docker rebuild

File: libafl_concolic/test/runtime_test/src/lib.rs
Comment (domenukk) [reply libafl_concolic/test/runtime_test/src/lib.rs:1]: Feature flag?

**Comment by domenukk** [comment]
@andreafioraldi want to take a look at this, too?

**Comment by andreafioraldi** [comment]
> @andreafioraldi want to take a look at this, too?

yup will do this afternoon in train

File: fuzzers/libfuzzer_stb_image_concolic/fuzzer/build.rs
Comment (andreafioraldi) [inline fuzzers/libfuzzer_stb_image_concolic/fuzzer/build.rs:98]: Consider to move this in a crate maybe?

**Comment by andreafioraldi** [comment]
LGTM, I have just one comment that can be addressed in the next PR

File: Dockerfile
Comment (julihoh) [reply Dockerfile:60]: Okay, I think I'm starting to understand what this is supposed to do. Before, I had just imitated what came before it.

I can't really remove the COPY commands here, because it would stop building. What should I do instead? I'm a bit lost...

**Comment by julihoh** [comment]
🎉

File: fuzzers/libfuzzer_stb_image_concolic/fuzzer/build.rs
Comment (julihoh) [reply fuzzers/libfuzzer_stb_image_concolic/fuzzer/build.rs:98]: I've put it on the todo list

File: Dockerfile
Comment (domenukk) [reply Dockerfile:60]: Copy over everything you need?

File: libafl_concolic/test/runtime_test/src/lib.rs
Comment (julihoh) [reply libafl_concolic/test/runtime_test/src/lib.rs:1]: also put this on the todo list :)

Comment (julihoh) [reply libafl_concolic/test/runtime_test/src/lib.rs:1]: Okay, so I checked this again. The problem is that runtime_test needs to be a separate crate, because it needs to produce the separate shared object file as compilation result. therefore, I think I can't put it in the `libafl_test` crate

**Pull request merged**
